orangecontrib/text/widgets/owtwitter.py

```diff
@@ -0,0 +1,275 @@
+
+        self.secret_line_edit = QtGui.QLineEdit(self)
+        self.secret_line_edit.editingFinished.connect(self.secret_changed)
+        self.secret_line_edit.setMinimumWidth(self.widgets_min_width)
+        layout.addRow('Secret', self.secret_line_edit)
+        self.set_key_list()
+        return layout
+
+    def sizeHint(self):
+        return QtCore.QSize(*self.size_hint)
+
+    @property
+    def key(self):
+        return self.recent_keys[0] if self.recent_keys else None
+
+    @property
+    def recent_keys(self):
+        return self.parent.recent_api_keys
+
+    def set_key_list(self):
+        self.key_combo.clear()
+        if self.recent_keys:
+            for key in self.recent_keys:
+                self.key_combo.addItem(key.consumer_key)
+            self.secret_line_edit.setText(self.recent_keys[0].consumer_secret)
+
+    def check_api_key(self):
+        key = twitter.Credentials(self.key_combo.currentText(),
+                                  self.secret_line_edit.text())
+        valid = key.check_credentials()
+        if valid:
+            self.api_key_check_label.setText("API key is valid.")
+            self.save_key(key)
+        else:
+            self.api_key_check_label.setText("API key not validated!")
+
+    def save_key(self, key):
+        if key in self.recent_keys:
+            self.key_selected(self.recent_keys.index(key))
+        else:
+            self.recent_keys.insert(0, key)
+            self.set_key_list()
+
+    def accept(self):
+        if self.key not in self.recent_keys:
+            self.check_api_key()
+        self.parent.update_api(self.key)
+        super().accept()
+
+    def key_selected(self, n):
+        key = self.recent_keys[n]
+        self.recent_keys.pop(n)
+        self.recent_keys.insert(0, key)
+        self.set_key_list()
+
+    def secret_changed(self):
+        pass
+
+
+class Output:
+    CORPUS = "Corpus"
+
+
+def api_required(func):
+    @functools.wraps(func)
+    def wrapper(self, *args, **kwargs):
+        if not self.api:
+            self.open_set_api_key_dialog_button.setFocus()
+            tip = 'Please provide a valid API key in order to get the data.'
+            QtGui.QToolTip.showText(QtGui.QCursor.pos(), tip)
+        else:
+            return func(self, *args, **kwargs)
+    return wrapper
+
+
+def word_list_required(func):
+    @functools.wraps(func)
+    def wrapper(self, *args, **kwargs):
+        if not self.query_words:
+            self.query_words_text_edit.setFocus()
+            tip = 'Please input query words.'
+            QtGui.QToolTip.showText(QtGui.QCursor.pos(), tip)
+        else:
+            return func(self, *args, **kwargs)
+    return wrapper
+
+
+class OWTwitter(OWWidget):
+    """
+    Attributes:
+        key (twitter.Credentials): Twitter API key/secret holder
+        api (twitter.TwitterAPI): Twitter API obj
+    """
+    name = "Twitter"
+    description = "Loads tweets from the Twitter API."
+    icon = "icons/Twitter.svg"
```

Comment:
Please set the priority here so that this widget will be higher in the Text Mining section (currently it is the last one). It should be among Corpus and NY Times since they all conceptually do the same thing.